tensordict/nn/storage.py

```diff
@@ -0,0 +1,346 @@
+        hash_module: torch.nn.Module,
+        aggregation_module: torch.nn.Module | None = None,
+    ):
+        self.in_keys = in_keys if isinstance(in_keys, List) else [in_keys]
+        self.out_keys = [index_key]
+
+        super().__init__()
+
+        self.aggregation_module = (
+            aggregation_module if aggregation_module else hash_module
+        )
+
+        self.hash_module = hash_module
+        self.index_key = index_key
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        hash_values = []
+
+        for k in self.in_keys:
+            hash_values.append(self.hash_module(tensordict[k]))
+
+        td_hash_value = self.aggregation_module(
+            torch.stack(
+                hash_values,
+                dim=-1,
+            ),
+        )
+
+        output = tensordict.clone(False)
+        output[self.index_key] = td_hash_value
+        return output
+
+
+class TensorDictStorage(
+    TensorDictModuleBase, TensorStorage[TensorDictModuleBase, TensorDictModuleBase]
+):
+    """A Storage for TensorDict.
+
+    This module resembles a storage. It takes a tensordict as its input and
+    returns another tensordict as output similar to TensorDictModuleBase. However,
+    it provides additional functionality like python map:
+
+    Examples:
+        >>> import torch
+        >>> from tensordict import TensorDict
+        >>> mlp = torch.nn.LazyLinear(out_features=64, bias=True)
+        >>> binary_to_decimal = BinaryToDecimal(
+        ...     num_bits=8, device="cpu", dtype=torch.int32, convert_to_binary=True
+        ... )
+        >>> query_module = QueryModule(
+        ...     in_keys=["key1", "key2"],
+        ...     index_key="index",
+        ...     hash_module=torch.nn.Sequential(mlp, binary_to_decimal),
+        ... )
+        >>> embedding_storage = FixedStorage(
+        ...     torch.nn.Embedding(num_embeddings=23, embedding_dim=1),
+        ...     lambda x: torch.nn.init.constant_(x, 0),
+        ... )
+        >>> tensor_dict_storage = TensorDictStorage(
+        ...     query_module=query_module,
+        ...     key_to_storage={"index": embedding_storage},
+        ... )
+        >>> index = TensorDict(
+        ...     {
+        ...         "key1": torch.Tensor([[-1], [1], [3], [-3]]),
+        ...         "key2": torch.Tensor([[0], [2], [4], [-4]]),
+        ...     },
+        ...     batch_size=(4,),
+        ... )
+        >>> value = TensorDict(
+        ...     {"index": torch.Tensor([[10], [20], [30], [40]])}, batch_size=(4,)
+        ... )
+        >>> tensor_dict_storage[index] = value
+        >>> tensor_dict_storage[index]
+        >>> assert torch.sum(tensor_dict_storage.contain(index)).item() == 4
+        >>> new_index = index.clone(True)
+        >>> new_index["key3"] = torch.Tensor([[4], [5], [6], [7]])
+        >>> retrieve_value = tensor_dict_storage[new_index]
+        >>> assert (retrieve_value["index"] == value["index"]).all()
```

Comment:
This example doesn't run. I think part of the problem is that aggregation_module in QueryModule falls back onto hash_module if not provided, but I fail to see the logic there...

Reply:
I have copied this from a unit test, but I did some changes afterward that may break the code. Let me copy / past the `test_storage` unit test again.